pkg/infoschema/infoschema_v2.go

```diff
@@ -276,16 +289,142 @@ func (isd *Data) deleteDB(dbInfo *model.DBInfo, schemaVersion int64) {
 	btreeSet(&isd.schemaID2Name, schemaIDName{schemaVersion: schemaVersion, id: dbInfo.ID, name: dbInfo.Name, tomb: true})
 }
 
-// GCOldVersion compacts btree nodes by removing items older than schema version.
-// exported for testing
-func (isd *Data) GCOldVersion(schemaVersion int64) (int, int64) {
-	maxv, ok := isd.byName.Load().Max()
-	if !ok {
-		return 0, 0
+type referredForeignKeysHelper struct {
+	start           referredForeignKeyItem
+	schemaVersion   int64
+	referredFKInfos []*model.ReferredFKInfo
+}
+
+func (h *referredForeignKeysHelper) onItem(item *referredForeignKeyItem) bool {
+	if item.dbName != h.start.dbName || item.tableName != h.start.tableName {
+		return false
 	}
 
-	var total int64
-	var deletes []*tableItem
+	if item.schemaVersion <= h.schemaVersion {
+		if !item.tomb { // If the item is a tomb record, all the foreign keys are deleted.
+			h.referredFKInfos = item.referredFKInfo
+		}
+		return false
+	}
+	return true
+}
+
+func (isd *Data) getTableReferredForeignKeys(schema, table string, schemaMetaVersion int64) []*model.ReferredFKInfo {
+	helper := referredForeignKeysHelper{
+		start:           referredForeignKeyItem{dbName: schema, tableName: table, schemaVersion: math.MaxInt64},
+		schemaVersion:   schemaMetaVersion,
+		referredFKInfos: make([]*model.ReferredFKInfo, 0),
+	}
+	isd.referredForeignKeys.Load().DescendLessOrEqual(&helper.start, helper.onItem)
+
+	sort.Slice(helper.referredFKInfos, func(i, j int) bool {
+		if helper.referredFKInfos[i].ChildSchema.L != helper.referredFKInfos[j].ChildSchema.L {
+			return helper.referredFKInfos[i].ChildSchema.L < helper.referredFKInfos[j].ChildSchema.L
+		}
+		if helper.referredFKInfos[i].ChildTable.L != helper.referredFKInfos[j].ChildTable.L {
+			return helper.referredFKInfos[i].ChildTable.L < helper.referredFKInfos[j].ChildTable.L
+		}
```

Comment:
Can we maintain the order at insert time, rather than do it every time when get?